contracts/psm/VaultManager.sol

```diff
@@ -0,0 +1,276 @@
+        require(amount > 0, "deposit amount cannot be zero");
+
+        // transfer token to this contract
+        IERC20(token).safeTransferFrom(msg.sender, address(this), amount);
+        _distribute(amount);
+
+        netDepositAmount += amount;
+        emit Deposit(amount);
+    }
+
+    function _distribute(uint256 amount) private {
+        uint256 remain = amount;
+        uint256 totalPoint = getTotalPoint();
+
+        // deposit token to adapters by adapter point
+        for (uint256 i = 0; i < adapters.length; i++) {
+            if (adapters[i].active) { // only active adapter can be used
+                //  adapterAmount = depositAmount * point / totalPoint
+                uint256 amt = Math.mulDiv(amount, adapters[i].point, totalPoint);
+                if (amt > 0) {
+                    IERC20(token).safeIncreaseAllowance(adapters[i].adapter, amt);
+                    IAdapter(adapters[i].adapter).deposit(amt);
+                    remain -= amt;
+                }
+            }
+        }
+
+        // if remain amount > 0, add to local token
+        if (remain > 0) {
+            localToken += remain;
+        }
+    }
+
+    function getTotalPoint() public view returns (uint256) {
+        uint256 totalPoint;
+        for (uint256 i = 0; i < adapters.length; i++) {
+            if (adapters[i].active) {
+                totalPoint += adapters[i].point;
+            }
+        }
+
+        return totalPoint;
+    }
+
+    /**
+      * @dev withdraw token from adapters, only PSM can call this function
+      * @param receiver receiver address
+      * @param amount withdraw amount
+      */
+    function withdraw(address receiver, uint256 amount) external onlyPSM {
+        require(amount > 0, "withdraw amount cannot be zero");
+
+        require(amount <= netDepositAmount, "withdraw amount exceeds net deposit amount");
+
+        netDepositAmount -= amount;
+
+        uint256 remain = amount;
+
+        // withdraw token from local first
+        if (localToken >= remain) {
+            IERC20(token).safeTransfer(receiver, remain);
+            localToken -= remain;
+            return;
+        } else {
+            IERC20(token).safeTransfer(receiver, localToken);
+            remain -= localToken;
+            localToken = 0;
+        }
+
+        // withdraw token from adapters
+        for (uint256 i = 0; i < adapters.length; i++) {
+            uint256 totalAvailableAmount = IAdapter(adapters[i].adapter).totalAvailableAmount();
+            if (totalAvailableAmount >= remain) {
+                IAdapter(adapters[i].adapter).withdraw(receiver, remain);
+                return;
+            } else {
+                IAdapter(adapters[i].adapter).withdraw(receiver, totalAvailableAmount);
+                remain -= totalAvailableAmount;
+            }
+        }
+
+        require(remain == 0, "not enough available balance");
+
+        emit Withdraw(receiver, amount);
```

Comment:
The function may return at line 168 so that the event may not get emitted.

Maybe we can revise the code to unify the logic.